llm/main.py

```diff
@@ -24,13 +26,30 @@ def build_logger(name, kwargs):
         raise ValueError(f'Not sure how to build logger: {name}')
 
 
-def build_callback(name, kwargs):
+def get_model_fwd_flops(cfg):
+    if cfg.model.name == 'mosaic_gpt':
+        # the number of paramters is approximately the number of multiply-accumulates (MAC) in the network
+        # each MAC has 2 FLOPs - we multiply by 2 ie 2 * n_param
+        # this gets us FLOPs / token
+        params_flops_per_token = 2 * cfg.n_params
+        params_flops_per_seq = params_flops_per_token * cfg.max_seq_len
+        # there are 2 FLOPS per mac; there is A=Q*K^T and out=A*V ops (ie mult by 2)
+        attn_flops_per_seq = cfg.model.n_layers * 2 * 2 * (cfg.model.d_model * (cfg.max_seq_len ** 2))
+        return params_flops_per_seq + attn_flops_per_seq
+    
+    return None
```

Comment:
Can we co-locate this with the model definition? My gut feeling is that it makes more sense there than in the training script. Like what if it lived on `ComposerModel` and then maybe the `SpeedMonitor` searches for it and calls if it exists?

Reply:
This function could also maybe be defaulted in the `ComposerModel` base class! Like `6 *  sum(p.numel() for p in model.parameters())`.

Reply:
`6 *  sum(p.numel() for p in model.parameters())`
only works when you have MLP heavy network (ie wont work for CV networks).
This is also FLOPs per token not sample.